nifi-nar-bundles/nifi-prometheus-bundle/nifi-prometheus-processors/src/main/java/org/apache/nifi/processors/prometheus/PrometheusRemoteWrite.java

```diff
@@ -0,0 +1,318 @@
+
+                WriteRequest writeRequest = WriteRequest.parseFrom(is);
+
+                for (Types.TimeSeries timeSeries: writeRequest.getTimeseriesList()) {
+                    List<MetricLabel> labelsList = new ArrayList<>();
+                    List<MetricSample> sampleList = new ArrayList<>();
+                    Metrics metrics = new Metrics();
+                    Gson gson = new Gson();
+
+                    for (Types.Label labelItem: timeSeries.getLabelsList()) {
+                        MetricLabel metricsLabel = new MetricLabel();
+                        metricsLabel.name = labelItem.getName();
+                        metricsLabel.value = labelItem.getValue();
+                        labelsList.add(metricsLabel);
+                    }
+                    for (Types.Sample sample: timeSeries.getSamplesList()) {
+                        MetricSample metricSample = new MetricSample();
+                        metricSample.sample = Double.toString(sample.getValue());
+                        metricSample.timestamp = Long.toString(sample.getTimestamp());
+                        sampleList.add(metricSample);
+                    }
+                    metrics.metricLabels= labelsList;
+                    metrics.metricSamples = sampleList;
+
+                    if (maxBatch == 0 || maxBatch == 1) {
+                        flowfile = session.create();
+                        session.write(flowfile, new OutputStreamCallback() {
+                            @Override
+                            public void process(OutputStream out) throws IOException {
+                                out.write(gson.toJson(metrics).getBytes());
+                                out.flush();
+                                out.close();
+                            }
+                        });
+                        getLogger().debug("Batch mode disabled written FlowFile: {}.", new Object[]{flowfile.getId()});
+                        commitFlowFile(flowfile, session, request);
+                    } else if (metricList.size() < maxBatch) {
+                        metricList.add(metrics);
+                    }
+
+                    if (metricList.size() == maxBatch) {
+                        flowfile = session.create();
+                        session.write(flowfile, new OutputStreamCallback() {
+                            @Override
+                            public void process(OutputStream out) throws IOException {
+                                out.write(gson.toJson(metricList).getBytes());
+                                out.flush();
+                                out.close();
+                            }
+                        });
+                        getLogger().debug("Batch mode enabled, writing {} metrics in FlowFile {}.",
+                                new Object[]{metricList.size(), flowfile.getId()});
+                        commitFlowFile(flowfile, session, request);
```

Comment:
Can `metricList.size()` equal 0 or 1? If so it appears the flowfile would get transferred twice (which would cause an error)